### PR TITLE
Kyc

### PR DESCRIPTION
**Pull request by shan-57blocks**
See: https://github.com/00labs/huma-dapp/pull/659

### CODE REVIEW

File: packages/huma-widget/src/components/Lend/supplyV2/components/PersonaEvaluation.tsx [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+  const [inquiryId, setInquiryId] = useState<string | undefined>(
+    localStorage.getItem(INQUIRY_ID) ?? undefined,
+  )
+  const [sessionToken, setSessionToken] = useState<string | undefined>()
+  const [verificationStatus, setVerificationStatus] = useState<
+    VerificationStatusResultV2 | undefined
+  >()
+  const [showPersonaClient, setShowPersonaClient] = useState<boolean>(false)
+  const isKYCCompletedRef = useRef<boolean>(false)
+  const isActionOnGoingRef = useRef<boolean>(false)
```
Comment (ljiatu) [inline packages/huma-widget/src/components/Lend/supplyV2/components/PersonaEvaluation.tsx:79]: nit: ongoing is one word, so should be `isActionOngoingRef`.

File: packages/huma-widget/src/components/Lend/supplyV2/components/PersonaEvaluation.tsx [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+    } else if (errorType === 'UserRejected') {
+      dispatch(
+        setError({
+          errorMessage: 'User has rejected the transaction.',
+        }),
+      )
+    } else if (errorType === 'Other') {
+      dispatch(
+        setError({
+          errorMessage: 'Something went wrong, please try again later.',
```
Comment (ljiatu) [inline packages/huma-widget/src/components/Lend/supplyV2/components/PersonaEvaluation.tsx:258]: nit: Should be two sentences, i.e. `Something went wrong. Please try again later.`

File: packages/huma-widget/src/components/Lend/supplyV2/components/PersonaEvaluation.tsx [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+
+  if (!loadingType) {
+    return (
+      <WrapperModal title={kycCopy.title}>
+        <Box css={styles.iconWrapper}>
+          <img src={ApproveLenderImg} alt='approve-lender' />
+        </Box>
+        <Box css={styles.description}>{kycCopy.description}</Box>
+        {Boolean(kycCopy.buttonText) && (
+          <BottomButton variant='contained' onClick={handleAction}>
```
Comment (ljiatu) [inline packages/huma-widget/src/components/Lend/supplyV2/components/PersonaEvaluation.tsx:345]: What happens if the button is clicked before `verificationStatus` is returned?

File: packages/huma-widget/src/components/Lend/supplyV2/components/PersonaEvaluation.tsx [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+    chainId,
+    changeTranche,
+    dispatch,
+    isDev,
+    isUniTranche,
+    poolInfo.juniorTrancheVault,
+    poolInfo.seniorTrancheVault,
+  ])
+
+  const checkVerificationStatus = useCallback(async () => {
```
Comment (ljiatu) [inline packages/huma-widget/src/components/Lend/supplyV2/components/PersonaEvaluation.tsx:120]: Should we check whether the lender has already been approved before checking their verification status?

File: packages/huma-widget/src/components/Lend/supplyV2/components/PersonaEvaluation.tsx [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+                await IdentityServiceV2.resumeVerification(
+                  account,
+                  chainId,
+                  isDev,
+                )
+              verificationStatus.status = resumeVerificationResult.status
+              setVerificationStatus(verificationStatus)
+              setSessionToken(resumeVerificationResult.sessionToken)
+            }
+            setKYCCopy(KYCCopies.verifyIdentity)
```
Comment (ljiatu) [inline packages/huma-widget/src/components/Lend/supplyV2/components/PersonaEvaluation.tsx:169]: If after the user has completed KYC the status is `PENDING` for an extended period of time, would we still show the wrong copy?

File: packages/huma-widget/src/components/Lend/supplyV2/components/PersonaEvaluation.tsx (hunk 4 above)
Comment (shan-57blocks) [reply packages/huma-widget/src/components/Lend/supplyV2/components/PersonaEvaluation.tsx:120]: This is checked before entering this component

File: packages/huma-widget/src/components/Lend/supplyV2/components/PersonaEvaluation.tsx (hunk 3 above)
Comment (shan-57blocks) [reply packages/huma-widget/src/components/Lend/supplyV2/components/PersonaEvaluation.tsx:345]: When calling checkVerificationStatus function, we show the loading without button.

File: packages/huma-shared/src/hooks/useAuthErrorHandling.ts [hunk 6]
```diff
@@ -6,6 +6,8 @@ import { JsonRpcProvider } from '@ethersproject/providers'
 import { useAsyncError } from './useAsyncError'
 import { AuthService } from '../services'
 
+type ErrorType = 'NotSignIn' | 'UserRejected' | 'Other' | undefined
```
[the remark this answers is not in the document]
Comment (shan-57blocks) [reply packages/huma-shared/src/hooks/useAuthErrorHandling.ts]: Updated

File: packages/huma-shared/src/services/IdentityServiceV2.ts [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+  personaInquiryId: string
+  isNotStarted: boolean
+}
+
+/**
+ * Get wallet's identity verification status.
+ *
+ * @param {string} walletAddress The wallet address.
+ * @param {string} pool The pool address.
+ * @param {number} chainId chain ID.
```
[the remark this answers is not in the document]
Comment (shan-57blocks) [reply packages/huma-shared/src/services/IdentityServiceV2.ts]: Updated

File: packages/huma-shared/src/services/IdentityServiceV2.ts [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+  walletAddress: string
+  status: IdentityVerificationStatusV2
+  personaInquiryId: string
+}
+
+/**
+ * Start wallet's verification process.
+ *
+ * @param {string} walletAddress The wallet address.
+ * @param {number} chainId chain ID.
```
[the remark this answers is not in the document]
Comment (shan-57blocks) [reply packages/huma-shared/src/services/IdentityServiceV2.ts]: Updated

File: packages/huma-shared/src/services/IdentityServiceV2.ts [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+
+  result.isNotStarted =
+    result.status === IdentityVerificationStatusV2.NOT_STARTED
+  return result
+}
+
+/**
+ * Object representing the response to the identity verification resume request.
+ * @typedef {Object} StartVerificationResultV2
+ * @property {string} walletAddress the wallet address to resume the verification.
```
[the remark this answers is not in the document]
Comment (shan-57blocks) [reply packages/huma-shared/src/services/IdentityServiceV2.ts]: Updated

File: packages/huma-shared/src/services/IdentityServiceV2.ts [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+  walletAddress: string
+  sessionToken: string
+  status: IdentityVerificationStatusV2
+}
+
+/**
+ * Resume wallet's verification process.
+ *
+ * @param {string} walletAddress The wallet address.
+ * @param {number} chainId chain ID.
```
[the remark this answers is not in the document]
Comment (shan-57blocks) [reply packages/huma-shared/src/services/IdentityServiceV2.ts]: Updated

File: packages/huma-shared/src/services/IdentityServiceV2.ts [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+  )
+
+  return result
+}
+
+/**
+ * Approve wallet as lender.
+ *
+ * @param {string} walletAddress The wallet address.
+ * @param {number} chainId chain ID.
```
[the remark this answers is not in the document]
Comment (shan-57blocks) [reply packages/huma-shared/src/services/IdentityServiceV2.ts]: Updated

File: packages/huma-shared/src/services/IdentityServiceV2.ts [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+ * @param {boolean} isDev Is dev environment or not.
+ * @returns {Promise<void>} Promise that returns void.
+ */
+const approveLender = async (
+  walletAddress: string,
+  chainId: number,
+  contractAddress: string,
+  isDev = false,
+): Promise<void> => {
+  const result = await requestPost<void>(
```
[the remark this answers is not in the document]
Comment (shan-57blocks) [reply packages/huma-shared/src/services/IdentityServiceV2.ts]: Updated

File: packages/huma-shared/src/services/IdentityServiceV2.ts [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+ * @param {number} chainId chain ID.
+ * @param {boolean} isDev Is dev environment or not.
+ * @returns {Promise<ResumeVerificationResultV2>} Promise that returns the start verification result.
+ */
+const resumeVerification = async (
+  walletAddress: string,
+  chainId: number,
+  isDev = false,
+): Promise<ResumeVerificationResultV2> => {
+  const result = await requestPost<ResumeVerificationResultV2>(
```
[the remark this answers is not in the document]
Comment (shan-57blocks) [reply packages/huma-shared/src/services/IdentityServiceV2.ts]: Updated

File: packages/huma-shared/src/services/IdentityServiceV2.ts [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+ * @param {number} chainId chain ID.
+ * @param {boolean} isDev Is dev environment or not.
+ * @returns {Promise<VerificationStatusResultV2>} Promise that returns the start verification result.
+ */
+const startVerification = async (
+  walletAddress: string,
+  chainId: number,
+  isDev = false,
+): Promise<VerificationStatusResultV2> => {
+  const result = await requestPost<VerificationStatusResultV2>(
```
[the remark this answers is not in the document]
Comment (shan-57blocks) [reply packages/huma-shared/src/services/IdentityServiceV2.ts]: Updated

File: packages/huma-shared/src/services/IdentityServiceV2.ts [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+ * @param {number} chainId chain ID.
+ * @param {boolean} isDev Is dev environment or not.
+ * @returns {Promise<VerificationStatusResultV2>} Promise that returns the verification status result.
+ */
+const getVerificationStatusV2 = async (
+  walletAddress: string,
+  chainId: number,
+  isDev = false,
+): Promise<VerificationStatusResultV2> => {
+  const result = await requestGet<VerificationStatusResultV2>(
```
[the remark this answers is not in the document]
Comment (shan-57blocks) [reply packages/huma-shared/src/services/IdentityServiceV2.ts]: Updated

File: packages/huma-widget/src/components/Lend/supplyV2/components/PersonaEvaluation.tsx [hunk 16]
```diff
@@ -0,0 +1,359 @@
...
+    changeTranche,
+    dispatch,
+    isDev,
+    isUniTranche,
+    poolInfo.juniorTrancheVault,
+    poolInfo.seniorTrancheVault,
+  ])
+
+  const checkVerificationStatus = useCallback(async () => {
+    if (isActionOnGoingRef.current === true) {
```
[the remark this answers is not in the document]
Comment (shan-57blocks) [reply packages/huma-widget/src/components/Lend/supplyV2/components/PersonaEvaluation.tsx]: Updated

File: packages/huma-widget/src/components/Lend/supplyV2/components/PersonaEvaluation.tsx [hunk 17]
```diff
@@ -0,0 +1,359 @@
...
+  const { account, chainId } = useWeb3React()
+  const {
+    errorType,
+    setError: setAuthError,
+    isWalletOwnershipVerified,
+    isWalletOwnershipVerificationRequired,
+  } = useAuthErrorHandling(isDev)
+  const KYCCopies = poolInfo.KYC!.Persona!
+  const [loadingType, setLoadingType] = useState<
+    'verificationStatus' | 'startKYC' | 'approveLender'
```
[the remark this answers is not in the document]
Comment (shan-57blocks) [reply packages/huma-widget/src/components/Lend/supplyV2/components/PersonaEvaluation.tsx]: Updated

File: packages/huma-widget/src/components/Lend/supplyV2/components/PersonaEvaluation.tsx [hunk 18]
```diff
@@ -0,0 +1,359 @@
...
+  >()
+  const [kycCopy, setKYCCopy] = useState<KYCCopy>(KYCCopies.verifyIdentity)
+  const [inquiryId, setInquiryId] = useState<string | undefined>(
+    localStorage.getItem(INQUIRY_ID) ?? undefined,
```
[the remark this answers is not in the document]
Comment (shan-57blocks) [reply packages/huma-widget/src/components/Lend/supplyV2/components/PersonaEvaluation.tsx]: Updated

File: packages/huma-widget/src/components/Lend/supplyV2/components/PersonaEvaluation.tsx (hunk 1 above)
Comment (shan-57blocks) [reply packages/huma-widget/src/components/Lend/supplyV2/components/PersonaEvaluation.tsx]: Updated

File: packages/huma-widget/src/components/Lend/supplyV2/components/PersonaEvaluation.tsx (hunk 2 above)
Comment (shan-57blocks) [reply packages/huma-widget/src/components/Lend/supplyV2/components/PersonaEvaluation.tsx]: Updated

File: packages/huma-widget/src/components/Lend/supplyV2/components/PersonaEvaluation.tsx (hunk 5 above)
Comment (shan-57blocks) [reply packages/huma-widget/src/components/Lend/supplyV2/components/PersonaEvaluation.tsx:163]: Updated

**Review by ljiatu — APPROVED** [review]

**Pull request merged**